Add TruLens third-party scorer integration

debu-sinha opened the pull request:
### Related Issues/PRs

Partial fix for #19062
Consolidates #19328 (agent trace scorers)

### What changes are proposed in this pull request?

Implements TruLens feedback functions as MLflow GenAI scorers for evaluating LLM applications. This enables users to leverage TruLens' evaluation capabilities directly within MLflow's evaluation framework.

**RAG Evaluation Scorers:**

| Scorer | Description |
|--------|-------------|
| `Groundedness` | Evaluates if outputs are grounded in context |
| `ContextRelevance` | Evaluates context relevance to query |
| `AnswerRelevance` | Evaluates answer relevance to query |
| `Coherence` | Evaluates logical flow of outputs |

**Agent Trace Evaluation Scorers:**

| Scorer | Description |
|--------|-------------|
| `LogicalConsistency` | Evaluates reasoning quality of agent traces |
| `ExecutionEfficiency` | Evaluates execution efficiency |
| `PlanAdherence` | Evaluates plan adherence |
| `PlanQuality` | Evaluates plan quality |
| `ToolSelection` | Evaluates tool selection quality |
| `ToolCalling` | Evaluates tool calling quality |

Based on TruLens' benchmarked goal-plan-action alignment evaluations achieving 95% error coverage against TRAIL.

---

## Usage Examples

### RAG Evaluation

```python
from mlflow.genai.scorers.trulens import Groundedness, ContextRelevance

# Direct scorer call
scorer = Groundedness(model="openai:/gpt-4o-mini")
feedback = scorer(trace=trace)

# With mlflow.genai.evaluate()
results = mlflow.genai.evaluate(
    data=eval_data,
    scorers=[
        Groundedness(model="openai:/gpt-4o-mini"),
        ContextRelevance(model="openai:/gpt-4o-mini"),
    ],
)
```

### Agent Trace Evaluation

```python
from mlflow.genai.scorers.trulens import LogicalConsistency, PlanAdherence

# Evaluate agent traces
traces = mlflow.search_traces(experiment_ids=["1"])
results = mlflow.genai.evaluate(
    data=traces,
    scorers=[
        LogicalConsistency(model="openai:/gpt-4o-mini"),
        PlanAdherence(model="openai:/gpt-4o-mini"),
    ],
)
```

---

**Model Support:**
- OpenAI: `model="openai:/gpt-4o-mini"`
- Databricks managed judge: `model="databricks"`
- Databricks serving endpoint: `model="databricks:/endpoint-name"`
- Any LiteLLM provider: `model="anthropic:/claude-3"`, `model="bedrock:/model-id"`, etc.

### How is this PR tested?

- [x] 46 unit tests (all passing)
- [x] Manual integration tests with real OpenAI API calls

---

## Real API Integration Test Results

### RAG Scorers (with OpenAI gpt-4o-mini)

```
1. Groundedness
   Input: "The Eiffel Tower is 330 meters tall and located in Paris."
   Context: "The Eiffel Tower stands at 330 meters in Paris, France."
   Value: yes
   Score: 1.0
   Threshold: 0.5

2. ContextRelevance
   Query: "What is the height of the Eiffel Tower?"
   Context: "The Eiffel Tower is 330 meters tall. It was built in 1889."
   Value: yes
   Score: 1.0

3. AnswerRelevance
   Query: "What is the capital of France?"
   Response: "Paris is the capital of France, known for the Eiffel Tower."
   Value: yes
   Score: 1.0

4. Coherence
   Text: "MLflow is an open-source platform. It manages the ML lifecycle..."
   Value: yes
   Score: 0.67
```

### Agent Trace Scorers (with OpenAI gpt-4o-mini)

```
Agent Trace: research_agent with planning → doc_search → summarize spans

5. LogicalConsistency
   Value: 1.0

6. ExecutionEfficiency
   Value: 1.0

7. PlanAdherence
   Value: 1.0

8. PlanQuality
   Value: 0.67

9. ToolSelection
   Value: 1.0

10. ToolCalling
    Value: 1.0
```

### Batch Evaluation with mlflow.genai.evaluate()

```python
# RAG scorers batch evaluation
results = mlflow.genai.evaluate(data=eval_data, scorers=[...])

Metrics:
   AnswerRelevance/mean: 1.00
   Coherence/mean: 1.00
   ContextRelevance/mean: 1.00
   Groundedness/mean: 1.00

# Agent trace scorers batch evaluation
results = mlflow.genai.evaluate(data=[trace], scorers=[...])

Metrics:
   execution_efficiency/mean: 1.00
   logical_consistency/mean: 1.00
   plan_adherence/mean: 1.00
```

---

### Does this PR require documentation update?

- [x] Yes. I've updated:
  - [x] API docstrings with examples

### Release Notes

#### Is this a user-facing change?

- [x] Yes

Added TruLens third-party scorer integration for MLflow GenAI evaluation:

**RAG Scorers:** `Groundedness`, `ContextRelevance`, `AnswerRelevance`, `Coherence`

**Agent Trace Scorers:** `LogicalConsistency`, `ExecutionEfficiency`, `PlanAdherence`, `PlanQuality`, `ToolSelection`, `ToolCalling`

Install dependencies:
```bash
pip install trulens trulens-providers-litellm
```

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [x] `area/evaluation`: MLflow model evaluation features

#### How should the PR be classified in the release notes?

- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes

#### Should this PR be included in the next patch release?

- [ ] Yes
- [x] No (this PR will be included in the next minor release)

debu-sinha committed:
Add TruLens third-party scorer integration

Implements TruLens feedback functions as MLflow GenAI scorers for
evaluating LLM applications within MLflow's evaluation framework.

TruLens scorers:
- Groundedness: Evaluates if outputs are grounded in context
- ContextRelevance: Evaluates context relevance to query
- AnswerRelevance: Evaluates answer relevance to query
- Coherence: Evaluates logical flow of outputs

Features:
- Multiple model providers: OpenAI, LiteLLM, Bedrock, Cortex
- Databricks managed judge support via call_chat_completions
- Databricks serving endpoint support
- Trace input extraction for RAG evaluation
- Configurable pass/fail threshold
- Consistent scorer interface returning Feedback objects

Partial fix for #19062

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Refactor TruLens module structure to match deepeval pattern

- Move TruLensScorer and metric classes from trulens.py to __init__.py
- Delete trulens.py (implementation now in __init__.py)

This matches the module structure used in mlflow/genai/scorers/deepeval/

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Add comprehensive unit tests for TruLens scorer integration

Adds test coverage for the TruLens scorer integration following
the same patterns established in the DeepEval tests:

- test_trulens.py: Core scorer functionality tests for all metric types
  (Groundedness, ContextRelevance, AnswerRelevance, Coherence)
- test_models.py: Provider creation tests for Databricks, OpenAI, LiteLLM
- test_registry.py: Metric registry lookup tests
- test_utils.py: Input mapping and rationale formatting tests

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha commented:
Added comprehensive unit tests for the TruLens scorer integration, following the same patterns from the DeepEval tests.

**Test Coverage:**
- `test_trulens.py`: Core scorer functionality for Groundedness, ContextRelevance, AnswerRelevance, and Coherence
- `test_models.py`: Provider creation tests (Databricks managed judge, serving endpoints, OpenAI, LiteLLM)
- `test_registry.py`: Metric registry lookup and configuration tests
- `test_utils.py`: Input mapping utilities and rationale formatting

All 42 tests passing locally.

debu-sinha committed:
Merge branch 'master' into feature/trulens-scorers

debu-sinha committed:
Remove fixture docstring and add TruLens to typos allowlist

- Remove one-liner docstring from mock_trulens_dependencies fixture
  to align with Phoenix PR review feedback
- Add 'Tru' to typos extend-words for TruLens library name

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Merge remote-tracking branch 'origin/master' into feature/trulens-scorers

debu-sinha committed:
Add TruLens dependencies to CI workflow

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Merge branch 'master' into feature/trulens-scorers

debu-sinha committed:
Align TruLens patterns with Phoenix integration

- Remove file docstrings from registry.py and utils.py
- Use pytest.importorskip at module level (tests fail if trulens not installed)
- Reduce mocking in tests: use real providers, only mock feedback method calls
- Add test_trulens_scorer_provider_is_real_instance to verify real providers
- Fix lint issues: parameterize dict type, use walrus operator
- Flatten nested with blocks in tests

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Merge branch 'master' into feature/trulens-scorers

debu-sinha committed:
Replace pytest.importorskip with direct imports to fail on missing deps

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Merge branch 'master' into feature/trulens-scorers

debu-sinha committed:
Fix lint and test issues for TruLens scorer integration

- Convert class-based tests to function-based in test_utils.py (MLF0029)
- Use @pytest.mark.parametrize for test cases with similar patterns
- Use monkeypatch.setenv for OpenAI API key in test_create_trulens_provider_openai

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

smoorjani commented on mlflow/genai/scorers/trulens/__init__.py:
```diff
@@ -0,0 +1,293 @@
+_logger = logging.getLogger(__name__)
+
+# Threshold for determining pass/fail
+DEFAULT_THRESHOLD = 0.5
```
```suggestion
_DEFAULT_THRESHOLD = 0.5
```
nit: let's make this internal

smoorjani commented on mlflow/genai/scorers/trulens/__init__.py:
```diff
@@ -0,0 +1,293 @@
+            scorer = Groundedness(model="openai:/gpt-4")
+            feedback = scorer(
+                outputs="The Eiffel Tower is 330 meters tall.",
+                expectations={"context": "The Eiffel Tower stands at 330 meters."},
```
it's a bit odd to pass context via `expectations` since context is not a ground-truth - can we have this example use `trace` instead? same above on L162

smoorjani commented on mlflow/genai/scorers/trulens/__init__.py:
```diff
@@ -0,0 +1,293 @@
+    supported by the source material.
+
+    Args:
+        model: Model URI (e.g., "openai:/gpt-4", "databricks", "databricks:/endpoint")
```
can we use `@format_docstring(_MODEL_API_DOC)` for this as well? same for all scorers below

smoorjani commented on mlflow/genai/scorers/trulens/__init__.py:
```diff
@@ -0,0 +1,293 @@
+    """
+    Evaluates whether the model's response is relevant to the input query.
+
+    Measures how well the answer addresses what was asked.
```
nit: slightly redundant

debu-sinha replied:
Fixed - updated docstring to be more meaningful.

smoorjani commented on mlflow/genai/scorers/trulens/models.py:
```diff
@@ -0,0 +1,194 @@
+            if prompt is None:
+                prompt = ""
+
+            result = call_chat_completions(user_prompt=prompt, system_prompt="")
```
maybe we can use this helper: `serialize_messages_to_databricks_prompts`? this should do something very similar

smoorjani commented on mlflow/genai/scorers/trulens/utils.py:
```diff
@@ -0,0 +1,123 @@
+                context_str = "\n".join(contexts)
+
+    # Map to TruLens-specific argument names based on metric
+    if metric_name == "Groundedness":
```
I would prefer not having metric-specific logic in here. Can we potentially pass everything into a TruLens metric? Alternatively, let's encode this information into the registry, keep your config pattern from registry.py, return everything from this function(e.g., all fields like source, statement, question, context, etc), and in the TruLens base scorer, use the information from the registry to determine what fields to keep.

smoorjani commented on mlflow/genai/scorers/trulens/utils.py:
```diff
@@ -0,0 +1,123 @@
+    input_str = parse_inputs_to_str(inputs) if inputs is not None else ""
+    output_str = parse_outputs_to_str(outputs) if outputs is not None else ""
+
+    # Extract context from expectations or trace
```
nit: can we use the trace for this instead? we typically only expect ground truths in expectations.

smoorjani commented on mlflow/genai/scorers/trulens/utils.py:
```diff
@@ -0,0 +1,123 @@
+        elif isinstance(value, dict):
+            parts.append(f"{key}: {value}")
+        else:
+            parts.append(f"{key}: {value}")
```
```suggestion
    for key, value in reasons.items():
        value_str = '; '.join(str(v) for v in value) if isinstance(value, list) else str(value)
        parts.append(f"{key}: {value_str}")
```

smoorjani commented on tests/genai/scorers/trulens/test_models.py:
```diff
@@ -0,0 +1,51 @@
+        yield mock
+
+
+def test_create_trulens_provider_databricks(mock_call_chat_completions):
```
can we test calling them and assert the sub-method (e.g., call_chat_completions) was called, similar tot he Arize tests. same for the ones below

smoorjani commented on tests/genai/scorers/trulens/test_registry.py:
```diff
@@ -0,0 +1,53 @@
+        get_feedback_method_name("InvalidMetric")
+
+
+def test_get_metric_config_groundedness():
```
let's update these tests after updating `registry.py` above, but generally aim to parameterize repetitive test patterns.

smoorjani commented on tests/genai/scorers/trulens/test_trulens.py:
```diff
@@ -0,0 +1,168 @@
+    )
+
+    assert result.value == CategoricalRating.NO
+    assert result.metadata["score"] == 0.3
```
let's assert over the entire result object so we know the exact output. same for tests below/above (e.g., include threshold)

debu-sinha replied:
Done. Integration tests use real scorer instances with only the provider's `_create_chat_completion` method mocked. Tests verify Feedback objects are created correctly and scores are returned.

smoorjani commented on tests/genai/scorers/trulens/test_utils.py:
```diff
@@ -0,0 +1,174 @@
+    assert format_trulens_rationale(reasons) == expected
+
+
+def test_format_trulens_rationale_multiple_reasons():
```
can we include these in the parameterized test above (`test_format_trulens_rationale`) - same for `test_format_trulens_rationale_dict_reason`

smoorjani reviewed (changes requested):
Thanks for writing this integration! Directionally, looks mostly good to me! Left comments to address, LMK if you have questions.

debu-sinha committed:
Merge master into feature/trulens-scorers

Resolve conflict in CI workflow by combining Phoenix and TruLens deps.
Switch to trulens-providers-litellm per reviewer feedback.

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Address smoorjani review feedback on TruLens scorers

- Make DEFAULT_THRESHOLD internal with _ prefix
- Add **kwargs support to pass to TruLens providers
- Simplify to use LiteLLM for all non-Databricks providers
- Update docstring examples to use trace instead of expectations
- Add @format_docstring decorator to all scorer classes
- Simplify registry by removing unused get_metric_config
- Simplify rationale formatting in utils
- Update tests to use LiteLLM and assert sub-method calls
- Fix pip install message to reference trulens-providers-litellm

Signed-off-by: Debu Sinha <debu.sinha@example.com>
Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Fix TruLens tests to skip on LiteLLM instrumentation bug

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha commented:
@smoorjani Thanks for the detailed feedback. I've addressed your comments:

**Changes Made:**

1. **`_DEFAULT_THRESHOLD`** - Made internal with `_` prefix

2. **`**kwargs` support** - Added to `TruLensScorer`, `get_scorer()`, and `create_trulens_provider()` to pass additional args to providers

3. **LiteLLM for all providers** - Simplified `models.py` to use LiteLLM for all non-Databricks providers (removed individual OpenAI/Bedrock/Cortex cases)

4. **Updated examples to use `trace`** - All docstring examples now use `scorer(trace=trace)` instead of expectations

5. **`@format_docstring(_MODEL_API_DOC)`** - Added to all scorer classes

6. **Simplified registry** - Removed unused `get_metric_config()`, simplified mapping to `dict[str, str]`

7. **Simplified rationale formatting** - Cleaned up `format_trulens_rationale()` per your suggestion

8. **Fixed pip install message** - Now references `trulens-providers-litellm`

9. **Tests updated**:
   - Assert sub-method calls for Databricks providers
   - Updated to use LiteLLM instead of OpenAI provider
   - Registry tests simplified after removing `get_metric_config`

**Test Results:**
```
33 passed, 2 skipped, 1 warning in 8.76s
```

The 2 skipped tests are due to a TruLens LiteLLM instrumentation bug (`AttributeError: 'CallTypes' object has no attribute '__name__'`) - this is an upstream issue in TruLens.

**Note:** I kept the metric-specific argument mapping in `utils.py` for now as each TruLens feedback function expects different argument names (`source`/`statement` for Groundedness vs `question`/`context` for ContextRelevance). Happy to refactor to registry-based approach if you have a preferred pattern.

debu-sinha commented:
@smoorjani I've addressed all the feedback in the previous commits. Here's a summary:

**Addressed:**
- `_DEFAULT_THRESHOLD` - made internal with underscore prefix
- Added `**kwargs` support to pass to TruLens classes
- Changed examples to use `trace` instead of `expectations`
- Applied `@format_docstring(_MODEL_API_DOC)` to all scorers
- Using LiteLLM for all non-Databricks providers (simplified provider handling)
- Updated pip install message
- Simplified registry by removing unused `get_metric_config`
- Simplified rationale formatting

**Also fixed:**
- Discovered and fixed a TruLens LiteLLM instrumentation bug that was causing `AttributeError` - filed issue #2327 and submitted fix PR truera/trulens#2328 (approved, pending merge)
- Added skip logic in tests as temporary workaround until TruLens fix is released

Could you please re-review when you have a chance? All tests pass locally.

debu-sinha commented:
Re: unresolved threads - I'll address the remaining items:

1. **Message handling in `_create_chat_completion`** - Will check if we should use `serialize_messages_to_databricks_prompts` helper
2. **`prompt=None` logic** - Will verify the fallback logic is correct

Looking at this now.

debu-sinha committed:
Improve message handling in TruLens Databricks providers

- Properly extract system prompts from messages for managed judge
- Fix prompt/message handling logic to avoid overwrites
- Add clear comments explaining the conversion flow

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha commented:
Pushed update addressing the message handling feedback:
- Properly extracts system prompts from messages for managed judge (uses separate `system_prompt` param)
- Fixed prompt/message handling logic to avoid the overwrite issue
- Added clear comments explaining the conversion flow

All 35 tests pass.

debu-sinha committed:
Remove single-line docstrings from TruLens models

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Address all review feedback for TruLens scorer integration

- Add parse_third_party_model_uri and serialize_chat_messages_to_prompts
  helpers to scorer_utils.py for shared functionality across integrations
- Refactor models.py to use shared helpers for message serialization
  and model URI parsing
- Move metric-specific argument mapping from utils.py to registry.py
  using TruLensMetricConfig dataclass
- Update utils.py to use build_trulens_args from registry
- Update test_utils.py to use real Trace objects instead of mocks
- Update test_trulens.py to assert entire result object including
  threshold and metadata
- Update test_registry.py for new registry structure with arg mappings

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Remove redundant comments

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Add TruLens agent trace scorers to integration

Add 6 new agent trace evaluation scorers:
- LogicalConsistencyScorer: Evaluates reasoning quality
- ExecutionEfficiencyScorer: Evaluates execution efficiency
- PlanAdherenceScorer: Evaluates plan adherence
- PlanQualityScorer: Evaluates plan quality
- ToolSelectionScorer: Evaluates tool selection
- ToolCallingScorer: Evaluates tool calling quality

These scorers analyze agent execution traces for goal-plan-action alignment,
based on TruLens' benchmarked evaluations achieving 95% error coverage.

Consolidates PR #19328 into this PR per reviewer feedback.

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Merge branch 'master' into feature/trulens-scorers

debu-sinha committed:
Fix formatting in TruLens models.py

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Simplify TruLens registry per maintainer feedback

- Remove TruLensMetricConfig dataclass
- Simplify registry to direct metric->method mapping
- Move arg_mapping logic to utils.py
- Update tests accordingly

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Merge branch 'master' into feature/trulens-scorers

debu-sinha commented on mlflow/genai/scorers/trulens/models.py:
```diff
@@ -0,0 +1,194 @@
+                        "OpenAI provider requires 'trulens-providers-openai'. "
+                        "Install it with: pip install trulens-providers-openai"
+                    )
+            case "litellm":
```
Done - now using LiteLLM for all non-Databricks providers. Updated install instructions to `trulens-providers-litellm`.

smoorjani commented on mlflow/genai/scorers/scorer_utils.py:
```diff
@@ -241,3 +241,70 @@ def get_tool_call_signature(call: "FunctionCall", include_arguments: bool) -> st
     return call.name
+
+
+def parse_third_party_model_uri(model_uri: str) -> tuple[str, str | None]:
```
there's already existing helpers like `_parse_model_uri` in the codebase - can we use that instead? can you also replace the usage in other 3p integrations like [here](https://sourcegraph.prod.databricks-corp.com/mlflow/mlflow/-/blob/mlflow/genai/scorers/deepeval/models.py?L75) - same for ragas and arize phoenix

smoorjani commented on mlflow/genai/scorers/trulens/__init__.py:
```diff
@@ -0,0 +1,289 @@
+_DEFAULT_THRESHOLD = 0.5
+
+
+@experimental(version="3.9.0")
```
I think we missed the release candidate for 3.9.0 - let's set this for 3.10.0rc0 (Scheduled in early feb)

smoorjani commented on mlflow/genai/scorers/trulens/agent_trace.py:
```diff
@@ -0,0 +1,264 @@
+        try:
+            trace_str = self._get_trace_string(trace)
+            feedback_method = getattr(self._provider, self._method_name)
+            score, reasons = feedback_method(trace=trace_str)
```
have we manually tested that this works? the MLflow trace/jsonified trace may be different than what Trulens expects

smoorjani commented on mlflow/genai/scorers/trulens/agent_trace.py:
```diff
@@ -0,0 +1,264 @@
+_logger = logging.getLogger(__name__)
+
+
+class _AgentTraceScorerBase(Scorer):
```
can this extend the other TruLens scorer? correct me if I'm wrong, but the only thing changing in the `__call__` is what we're passing into the `feedback_method`, so we can pull that into a separate method (e.g., `_get_feedback_method_params`) and override that in `class TruLensAgentScorer(TruLensScorer)`

smoorjani commented on tests/genai/scorers/trulens/test_agent_trace.py:
```diff
@@ -0,0 +1,181 @@
+from unittest.mock import Mock, patch
```
same as above to move this into `tests/genai/scorers/trulens/scorers/test_agent_trace.py`

smoorjani reviewed (changes requested):
almost there, mostly LGTM! mostly some style stuff but functionally looks great!

debu-sinha committed:
Address review feedback for TruLens integration

- Update @experimental version from 3.9.0 to 3.10.0
- Restructure agent_trace.py to scorers/ subfolder
- Rename agent scorer classes (remove "Scorer" suffix)
- Rename base class to TruLensAgentScorer
- Simplify models.py: remove serving endpoint path, use LiteLLM for all
- Add agent metrics to registry
- Rename format_trulens_rationale to format_rationale
- Add metadata to error cases
- Update tests per review feedback
- Remove single-line docstrings from tests
- Move importlib import to top level

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Address remaining review comments from smoorjani

- Move _ARG_MAPPING to registry.py and integrate with metric registry
- Add get_argument_mapping() and build_trulens_args() to registry
- Add validation warning when multiple context keys are provided
- Remove redundant test_scorers_available_in_trulens_namespace test
- Add unit tests for parse_third_party_model_uri and serialize_chat_messages_to_prompts

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Merge branch 'master' into feature/trulens-scorers

debu-sinha committed:
Merge branch 'master' into feature/trulens-scorers

smoorjani reviewed (changes requested):
pretty much almost there! left some nits and I think some changes may not have been pushed

debu-sinha committed:
Address review feedback: remove parse_third_party_model_uri

- Use inline model URI parsing in TruLens models.py (matching DeepEval/Phoenix pattern)
- Remove parse_third_party_model_uri from scorer_utils.py
- Remove corresponding tests from test_scorer_utils.py
- Add message_utils.py that was previously uncommitted

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Address remaining review feedback

- Update model examples from openai:/gpt-4 to openai:/gpt-5
- Remove "databricks" from OSS docstring examples
- Move serialize_chat_messages_to_prompts from scorer_utils.py to message_utils.py
- Update test_utils.py to use mlflow.start_span instead of constructing traces directly

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Simplify test_trulens_scorer parameterization

- Merge test_trulens_scorer_fail into parameterized test
- Remove redundant metric_name parameter (same as scorer_name)

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Use _parse_model_uri across all 4 integrations, consolidate message_utils

- Use _parse_model_uri from mlflow/metrics/genai/model_utils.py in TruLens, DeepEval, Phoenix, and Ragas models
- Move serialize_messages_to_databricks_prompts to mlflow/genai/utils/message_utils.py
- Update imports in databricks_managed_judge_adapter.py and simulator.py
- Update tests to match new error message format from _parse_model_uri

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Add unit tests for message_utils.py

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Add assertion to verify call_chat_completions is called in Databricks provider test

Test now validates that the Databricks managed judge provider uses call_chat_completions
as expected when _create_chat_completion is invoked.

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha commented:
All review comments have been addressed:

1. **Model examples updated** - Changed from `openai:/gpt-4` to `openai:/gpt-5` across all docstrings
2. **Removed "databricks" from OSS docs** - Examples now use `model="openai:/gpt-5"` consistently  
3. **Using `_parse_model_uri` helper** - Replaced inline parsing with shared helper from `mlflow/metrics/genai/model_utils.py` across TruLens, DeepEval, Phoenix, and Ragas integrations
4. **Consolidated message utilities** - Moved `serialize_messages_to_databricks_prompts` and `serialize_chat_messages_to_prompts` to `mlflow/genai/utils/message_utils.py` and updated all imports
5. **Added tests for message utils** - Created `tests/genai/utils/test_message_utils.py` with 8 unit tests
6. **Test improvements**:
   - Merged pass/fail test cases into parameterized test
   - Simplified trace creation using `mlflow.start_span()`
   - Added assertion to verify `call_chat_completions` is called in Databricks provider test

All TruLens tests pass (46 tests). Ready for re-review.

smoorjani commented on mlflow/genai/utils/message_utils.py:
```diff
@@ -0,0 +1,73 @@
+from typing import Any
+
+
+def serialize_messages_to_databricks_prompts(
```
can we merge these two functions? they look quite similar

smoorjani reviewed (approved):
left a few nits to address before merging, otherwise looks great!

debu-sinha committed:
Merge branch 'master' into feature/trulens-scorers

AveshCSingh reviewed (approved):
I mostly defer to Samraj's thourough review --the implementation + validations described in the PR looks reasonable. Left one small comment inline. 

One thing we should consider is whether to hook up 3p scorers with the MLflow AI Gateway. This does not block the PR merge though, and is a potential future improvement. cc @BenWilson2 @B-Step62

debu-sinha committed:
Address review feedback: merge message serialization functions

- Merge serialize_messages_to_databricks_prompts and serialize_chat_messages_to_prompts
  into a unified serialize_messages_to_prompts function that handles both Message
  objects and dicts
- Add backwards compatibility aliases for existing imports
- Refactor test_message_utils.py: remove test class, use real ChatMessage objects
  instead of mocks, parameterize tests
- Remove misplaced serialization tests from test_scorer_utils.py (now covered in
  test_message_utils.py)

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Merge branch 'master' into feature/trulens-scorers

debu-sinha committed:
Fix test expectations for model URI format validation

Update tests to expect MlflowException when model URI lacks required slash.
The _parse_model_uri function requires format provider:/model, not provider:model.

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Merge branch 'master' into feature/trulens-scorers

debu-sinha committed:
Merge origin/master - include all third-party scorer dependencies

Resolve CI workflow conflict by including all packages:
- deepeval, ragas, arize-phoenix-evals (existing)
- trulens, trulens-providers-litellm (TruLens PR)
- guardrails-ai (from master)

Signed-off-by: debu-sinha <debusinha2009@gmail.com>
